File: swift/187_RepeatedDNASequences.swift
class Solution {
    func findRepeatedDnaSequences(_ s: String) -> [String] {
        if s.count <= 10 { return [] }
        var sequences: [Substring : Int] = [Substring : Int]()
        var trail = s.startIndex, start = s.index(s.startIndex, offsetBy: 9)
        while start < s.endIndex {
            var substr = s[trail...start]
            sequences[substr, default: 0] += 1
            
            s.formIndex(after: &start)
            s.formIndex(after: &trail)
        }
        
        
        var result: [String] = [String]()
        sequences.forEach{
            k, v in
            if v > 1 { result.append(String(k)) }
        }
        
        return result
    }
}

File: java/187_RepeatedDNASequences.java
class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        HashMap<String, Integer> seqs = new HashMap<String, Integer>();
        int trail = 0;
        for (int i = 9; i < s.length(); ++i) {
            String substr = s.substring(trail, i+1);
            
            int n = seqs.getOrDefault(substr, 0);
            seqs.put(substr, n + 1);
            ++trail;
        }
        
        List<String> result = new ArrayList<String>();
        seqs.forEach((k, v) -> {
            if (v > 1)
                result.add(k);
        });
        return result;
    }
}
